crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+					Self::deposit_event(RawEvent::AuctionClosed(collection_id, token_id, AuctionClosureReason::VendorCancelled));
+				},
+				None => {},
+			}
+		}
+
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Transfer ownership of a token. modifies storage, does no verification, infallible.
+	fn transfer_ownership(
+		collection_id: &[u8],
+		token_id: T::TokenId,
+		current_owner: &T::AccountId,
+		new_owner: &T::AccountId,
+	) {
+		// Update token ownership
+		<CollectedTokens<T>>::mutate(collection_id, current_owner, |tokens| tokens.retain(|t| t != &token_id));
+		<TokenOwner<T>>::insert(collection_id, token_id, new_owner);
+		<CollectedTokens<T>>::append(collection_id, new_owner, token_id);
+	}
+	/// Remove a single direct listing and all it's metadata
+	fn remove_direct_listing(collection_id: &CollectionId, token_id: T::TokenId) {
+		let listing_type = Listings::<T>::take(collection_id, token_id);
+		ListingWinningBid::<T>::remove(collection_id, token_id);
+		if let Some(Listing::<T>::DirectSale(listing)) = listing_type {
+			ListingEndSchedule::<T>::mutate(listing.close, |listings| {
+				listings.retain(|l| l != &(collection_id.clone(), token_id));
+			});
+		}
+	}
+	/// Close all listings scheduled to close at this block `now`, ensuring payments and ownerships changes are made for winning bids
+	/// Metadata for listings will be removed from storage
+	/// Returns the number of listings removed
+	fn close_listings_at(now: T::BlockNumber) -> u32 {
+		let mut removed = 0_u32;
+		for (collection_id, token_id) in ListingEndSchedule::<T>::take(now).into_iter() {
+			match Listings::<T>::take(&collection_id, token_id) {
+				Some(Listing::DirectSale(_)) => {
+					Self::deposit_event(RawEvent::DirectSaleClosed(collection_id.clone(), token_id));
+				}
+				Some(Listing::Auction(listing)) => {
+					if let Some((winner, hammer_price)) = ListingWinningBid::<T>::take(&collection_id, token_id) {
+						if let Err(err) =
+							Self::settle_auction(&collection_id, token_id, &listing, &winner, hammer_price)
+						{
+							// auction settlement failed despite our prior validations.
+							// release winning bid tokens. listing metadadta is removed by now.
+							log!(error, "🃏 auction settlement failed: {:?}", err);
+							T::MultiCurrency::unreserve(&winner, Some(listing.payment_asset), hammer_price);
+
+							Self::deposit_event(RawEvent::AuctionClosed(
+								collection_id.clone(),
+								token_id,
+								AuctionClosureReason::SettlementFailed,
+							));
+						} else {
+							// auction settlement success
+							Self::deposit_event(RawEvent::AuctionSold(
+								collection_id.clone(),
+								token_id,
+								listing.payment_asset,
+								hammer_price,
+								winner,
+							));
+						}
+					} else {
+						// normal closure, no acceptable bids
+						Self::deposit_event(RawEvent::AuctionClosed(
+							collection_id.clone(),
+							token_id,
+							AuctionClosureReason::ExpiredNoBids,
+						));
+					}
+				}
+				_ => (),
```

Comment:
Please remove this line, so if in the future someone added another item to the Listing enum, they automatically know they should cover the case here as well.